### PR TITLE
Increase speed of name matching for paired end reads.

### PR DESCRIPTION
**Pull request by rhpvorderman**
I noticed that paired end reading incurs a heavy performance penalty
Baseline:
```
Benchmark #1: python dnaio_read_paired.py ~/test/big2.fastq
  Time (mean ± σ):      5.693 s ±  0.026 s    [User: 5.480 s, System: 0.213 s]
  Range (min … max):    5.650 s …  5.737 s    10 runs
```
Given that in single-end mode it runs in 2.1 seconds, this is quite heavy. 
So I disabled the namecheck to see what the difference would be:
```
Benchmark #1: python dnaio_read_paired.py ~/test/big2.fastq
  Time (mean ± σ):      4.491 s ±  0.025 s    [User: 4.268 s, System: 0.223 s]
  Range (min … max):    4.443 s …  4.538 s    10 runs
```
So that is roughly 1.3 seconds extra. Just for the name checking. That is rather suboptimal. We can parse the entire file again in that time.  (NOTE that this PR originates from main, not the pending speed improvements.)

So this PR improves the speed of the name comparison. 

+ It checks if the strings are stored as 1-byte data. (They should be!). It throws a ValueError if not.
+ Then it retrieves a char pointer to the unicode buffer.
+ Since it is 1-byte data we know that space == 32 and that tab == 9. Use memchr to look for both space and tab.
+ Compare the unicode buffers. We use the length until the first space or tab character, or the entire length if there is no whitespace.

I also did a naive implementation where I checked every character individually for space and tab and also compared the header1 and header2 character, but that was significantly slower than using memchr on the entire string twice to find tab or space and then using memcmp on the entire string.

The result:
```
Benchmark #1: python dnaio_read_paired.py ~/test/big2.fastq
  Time (mean ± σ):      4.786 s ±  0.027 s    [User: 4.570 s, System: 0.215 s]
  Range (min … max):    4.733 s …  4.820 s    10 runs
```

So we now only spent 300 ms for the name checking. Far more acceptable.

EDIT: I just realized that we can also use this PyUnicode_1BYTE_DATA macro to make the writing of FASTQ records quite a lot faster. But I will implement that after this PR is merged.

### CODE REVIEW

File: src/dnaio/_core.pyx [hunk 1]
```diff
@@ -380,12 +384,56 @@ def record_names_match(header1: str, header2: str):
...
+        space_at = length
+    else:
+        space_at = space_ptr - str_ptr
+    # Look if there is a tab before the space
+    tab_ptr = <char *>memchr(str_ptr, 9, space_at)
+    if tab_ptr == NULL:
+        return space_at
+    if tab_ptr < space_ptr:
+        return tab_ptr - str_ptr
+    return space_at
```
Comment (marcelm) [inline src/dnaio/_core.pyx:439]: Can you please add a test for `record_names_match` that tests whether the above works with 1) no whitespace, 2) space before tab, 3) tab before space? I think it does, but this would guard against anyone trying to "simplify" the above incorrectly.

**Review by marcelm — COMMENTED** [review]
Thanks once again! This is quite relevant because so much data is paired-end. Just a couple of comments.

File: src/dnaio/_core.pyx [hunk 2]
```diff
@@ -380,12 +384,56 @@ def record_names_match(header1: str, header2: str):
...
+        PyUnicode_KIND(header2) != PyUnicode_1BYTE_KIND):
+        raise ValueError("Strings are not valid latin-1 encoded data.")
+    # No encoding, we can compare the unicode data directly. Provided it is
+    # in 1-byte encoding, so we can find the spaces and tabs easily.
+    cdef char * header1chars = PyUnicode_1BYTE_DATA(header1)
+    cdef char * header2chars = PyUnicode_1BYTE_DATA(header2)
+    return record_name_bytes_match(header1chars, header2chars, len(header1), len(header2))
+
+
+cdef object record_name_bytes_match(char *header1chars, char* header2chars, Py_ssize_t header1length, Py_ssize_t header2length):
```
[the remark this answers is not in the document]
Comment (rhpvorderman) [reply src/dnaio/_core.pyx:397]: yes. But for some reason I do not understand it leads to a much slower run in the benchmarks. It negates most of the speed benefit. I have no idea why. This might be due to alignment issues.

**Comment by rhpvorderman** [comment]
@marcelm Thanks for the review. All comments have been addressed.

File: src/dnaio/_core.pyx [hunk 3]
```diff
@@ -380,12 +384,56 @@ def record_names_match(header1: str, header2: str):
...
-    name2 = header2.split()[0]
-    if name1 and name2 and name1[-1] in '123' and name2[-1] in '123':
-        name1 = name1[:-1]
-        name2 = name2[:-1]
-    return name1 == name2
+    if (PyUnicode_KIND(header1) != PyUnicode_1BYTE_KIND or
+        PyUnicode_KIND(header2) != PyUnicode_1BYTE_KIND):
+        raise ValueError("Strings are not valid latin-1 encoded data.")
+    # No encoding, we can compare the unicode data directly. Provided it is
+    # in 1-byte encoding, so we can find the spaces and tabs easily.
```
[the remark this answers is not in the document]
Comment (rhpvorderman) [reply src/dnaio/_core.pyx]: Clarified to not calling '.encode' functions.

File: src/dnaio/_core.pyx (hunk 2 above)
[the remark this answers is not in the document]
Comment (rhpvorderman) [reply src/dnaio/_core.pyx]: Apparantly it does not matter in another context. Changed.

**Comment by rhpvorderman** [comment]
I did some searching in the GNU Libc standard library and I found `strcspn` which looks trough a string for a range of characters or the first NULL byte. 
While this is less efficiently implemented than memchr, it has about the same efficiency when looking for two characters. I could not measure a difference between my custom memchr code and this code. (Maybe the difference is too small too measure?). Regardless it saves a lot of code by using a libc function instead of rolling my own. strncp does not require length, but this is not needed since every python string is \00 terminated anyway.

The advantage is that each character in the string is only checked twice. Once for whitespace and once for the memcmp. 

I do want to note that I know that strncmp exists, but that is different from memcmp in that it also checks for a "\0" byte. Therefore it uses a check for null bytes every iteration. Memcmp does not do this 0 check and is therefore faster.

**Comment by marcelm** [comment]
Perfect, this looks good now!
